include/nbl/builtin/hlsl/blit/blit.hlsl

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BLIT_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BLIT_INCLUDED_
+
+
+#include <nbl/builtin/hlsl/multi_dimensional_array_addressing/multi_dimensional_array_addressing.hlsl>
+#include <nbl/builtin/hlsl/blit/parameters.hlsl>
+#include <nbl/builtin/hlsl/blit/common.hlsl>
+#include <nbl/builtin/hlsl/glsl_compat/core.hlsl>
+#include <nbl/builtin/hlsl/spirv_intrinsics/core.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace blit
+{
+
+// p in input space
+int32_t3 getMinKernelWindowCoord(NBL_CONST_REF_ARG(int32_t3) p, NBL_CONST_REF_ARG(float3) minSupport)
+{
+	return int32_t3(ceil(p - float3(0.5f, 0.5f, 0.5f) + minSupport));
+}
+
+template <
+	uint32_t SMemFloatsPerChannel,
+	uint32_t BlitOutChannelCount,
+	uint32_t BlitDimCount,
+	uint32_t AlphaBinCount,
+	uint32_t WorkGroupSize,
+	E_FORMAT SoftwareEncodeFormat,
+	typename InTexture,
+	typename OutTexture,
+	typename KernelWeights,
+	typename Statistics,
+	typename SharedAccessor>
+void blit(
+	NBL_CONST_REF_ARG(InTexture) inTex,
+	NBL_REF_ARG(OutTexture) outTex,
+	NBL_CONST_REF_ARG(KernelWeights) kernelWeights,
+	NBL_REF_ARG(Statistics) statistics,
+	NBL_CONST_REF_ARG(SharedAccessor) sharedAccessor,
+	NBL_CONST_REF_ARG(parameters_t) params,
+	uint32_t3 groupID,
+	uint32_t groupIndex)
+{
+	const uint32_t3 inDim = params.getInputImageDimensions();
+	const uint32_t3 outDim = params.getOutputImageDimensions();
+
+	const uint32_t3 windowDim = params.getWindowDimensions();
+	const uint32_t3 phaseCount = params.getPhaseCount();
+
+	const uint32_t3 outputTexelsPerWG = params.getOutputTexelsPerWG();
+
+
+	const float3 scale = params.fScale;
+	const float3 halfScale = scale * float3(0.5f, 0.5f, 0.5f);
+
+	const uint32_t3 minOutputPixel = groupID * outputTexelsPerWG;
+	const float3 minOutputPixelCenterOfWG = float3(minOutputPixel)*scale + halfScale;
+	const int32_t3 regionStartCoord = getMinKernelWindowCoord(minOutputPixelCenterOfWG, params.negativeSupport); // this can be negative, in which case HW sampler takes care of wrapping for us
```

Comment:
I'd wrap these in a `derived_parameters_t` struct and compute them inside a `static derived_parameters_t create(NBL_REF_ARG(parameters_t) prime)`

Reply:
this will also help you get rid of the free floating `getMinKernelWindowCoord` function